pkg/file/redundancy/getter/getter.go

```diff
@@ -0,0 +1,226 @@
+
+package getter
+
+import (
+	"context"
+	"io"
+	"sync"
+	"sync/atomic"
+	"time"
+
+	"github.com/ethersphere/bee/pkg/storage"
+	"github.com/ethersphere/bee/pkg/swarm"
+	"github.com/klauspost/reedsolomon"
+)
+
+// decoder is a private implementation of storage.Getter
+// if retrieves children of an intermediate chunk potentially using erasure decoding
+// it caches sibling chunks if erasure decoding started already
+type decoder struct {
+	fetcher    storage.Getter  // network retrieval interface to fetch chunks
+	putter     storage.Putter  // interface to local storage to save reconstructed chunks
+	addrs      []swarm.Address // all addresses of the intermediate chunk
+	inflight   []atomic.Bool   // locks to protect wait channels and RS buffer
+	cache      map[string]int  // map from chunk address shard position index
+	waits      []chan struct{} // wait channels for each chunk
+	rsbuf      [][]byte        // RS buffer of data + parity shards for erasure decoding
+	ready      chan struct{}   // signal channel for successful retrieval of shardCnt chunks
+	shardCnt   int             // number of data shards
+	parityCnt  int             // number of parity shards
+	wg         sync.WaitGroup  // wait group to wait for all goroutines to finish
+	mu         sync.Mutex      // mutex to protect buffer
+	fetchedCnt atomic.Int32    // count successful retrievals
+	cancel     func()          // cancel function for RS decoding
+	remove     func()          // callback to remove decoder from decoders cache
+}
+
+type Getter interface {
+	storage.Getter
+	io.Closer
+}
+
+// New returns a decoder object used tos retrieve children of an intermediate chunk
+func New(addrs []swarm.Address, shardCnt int, g storage.Getter, p storage.Putter, strategy Strategy, strict bool, fetchTimeout time.Duration, remove func()) Getter {
+	ctx, cancel := context.WithCancel(context.Background())
+	size := len(addrs)
+	if fetchTimeout == 0 {
+		fetchTimeout = 30 * time.Second
+	}
+	strategyTimeout := StrategyTimeout
+
+	rsg := &decoder{
+		fetcher:   g,
+		putter:    p,
+		addrs:     addrs,
+		inflight:  make([]atomic.Bool, size),
+		cache:     make(map[string]int, size),
+		waits:     make([]chan struct{}, shardCnt),
+		rsbuf:     make([][]byte, size),
+		ready:     make(chan struct{}, 1),
+		cancel:    cancel,
+		remove:    remove,
+		shardCnt:  shardCnt,
+		parityCnt: size - shardCnt,
+	}
+
+	// after init, cache and wait channels are immutable, need no locking
+	for i := 0; i < shardCnt; i++ {
+		rsg.cache[addrs[i].ByteString()] = i
+		rsg.waits[i] = make(chan struct{})
+	}
+
+	// prefetch chunks according to strategy
+	rsg.wg.Add(1)
+	go func() {
+		rsg.prefetch(ctx, strategy, strict, strategyTimeout, fetchTimeout)
+		rsg.wg.Done()
+	}()
+	return rsg
+}
+
+// Get will call parities and other sibling chunks if the chunk address cannot be retrieved
+// assumes it is called for data shards only
+func (g *decoder) Get(ctx context.Context, addr swarm.Address) (swarm.Chunk, error) {
```

Comment:
what this `Get` should ideally do is to wait for either the chunk channel `g.waits[i]` to finish OR wait for a signal for the recovery to finish.
there should also be an error channel that returns when a chunk is not fetchable AND unreconvery which would translate into a `storage.ErrNotFound`.

as it stands, this `GET` nevers properly returns an error, and simply waits for a context timeout in this case that recovery and/or fetch failed.